File: interview_prepartation_kit/trees/CHECK_BST.java
/* Hidden stub code will pass a root argument to the function below. Complete the function to solve the challenge. Hint: you may want to write one or more helper functions.  

The Node class is defined as follows:
    class Node {
        int data;
        Node left;
        Node right;
     }
*/
boolean checkBST(Node root) {
    return BST(root,Integer.MIN_VALUE,Integer.MAX_VALUE);
}
boolean BST(Node root,int min,int max)
{
    if(root==null)
    {
        return true;
    }
    if(root.data > min && root.data < max && BST(root.left,min,root.data) && BST(root.right,root.data,max))
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: Problem_solving/data_structures/Trees/CHECK_BST.cpp
/*The tree node has data, left child and right child 
class Node {
    int data;
    Node* left;
    Node* right;
};

*/
    Node *find(Node *root,int x)
    {
        Node *current=root;
        while(current)
        {   
             if(x > current->data)
             {  
                if(current->right->data==x)
                {
                    break;
                }
                else
                {
                current=current->right;
                }
             }
            else if(x < current->data)
             {
                if(current->left->data==x)
                {
                    break;
                }
                else
                {
                current=current->left;
                }
            }
        }
        return current;
    }
    Node *lca(Node *root, int v1,int v2) {
        Node *current;
        if(v1==v2)
        {
            current= find(root,v1);
        }
        else if(v1==root->data || v2==root->data)
        {
            return root;
        }
        else
        {   
            int max=v1>v2?v1:v2;
            int min=v1<v2?v1:v2;
            if(max>root->data && min<root->data)
            {
            current = root;
            }
            else if(max < root->data)
            {
                current=lca(root->left,max,min);
            }
            else if(min>root->data)
            {   
                current=lca(root->right,max,min);
            }
        }
        return current;
    }
